### PR TITLE
WIP: Fix x86 16 pcrel

### PR DESCRIPTION
**Pull request by ret2libc**
For now I'm using this capstone: https://github.com/ret2libc/capstone/tree/fix-x86-16-pcrel

I'm trying to upstream the patch (https://github.com/aquynh/capstone/pull/1225)

### CODE REVIEW

File: shlr/meson.build [hunk 1]
```diff
@@ -10,40 +10,26 @@ if not capstone_dep.found() or not get_option('use_sys_capstone')
...
 
-    # NOTE: when you update CS_TIP, also update it in shlr/Makefile
-    CS_TIP = '782ea67e17a391ca0d3faafdc365b335a1a8930a'
+    # NOTE: when you update CS_TIP or CS_BRA, also update them in shlr/Makefile
+    CS_TIP = ''
+    CS_BRA = 'fix-x86-16-pcrel'
 
-    message('Cloning capstone next branch, commit ' + CS_TIP + ', into ' + capstone_path)
-    git_cmd = 'clone -b next https://github.com/aquynh/capstone.git @0@'.format(capstone_path)
+    capstone_git_user = 'ret2libc'
```
Comment (XVilka) [inline shlr/meson.build]: hm?

**Comment by ret2libc** [comment]
TODO: fix parse color

**Comment by ret2libc** [comment]
@radare @XVilka 
I'm not really sure which one is better
1) `jmp 0x4c5`
2) `jmp 0xf000:0x4c5`
3) `jmp cs:0x4c5`

What do you think? This PR for the moment uses 2, but I think it might be easier and clear in the same way if we use 3.

**Comment by ret2libc** [comment]
We could even have a variable to change between 2 and 3.

**Comment by ret2libc** [comment]
I'm not sure the change of colors is really a problem. The fact was that before the r_parse_filter function was *sometimes* executed after the color was applied. In particular this resulted in strings/flags being colored in the same way as constants. Since I've made sure that r_parse_filter is called *before*color is applied, now strings/flags have a different color. It seems acceptable to me, but if it's not, we probably need to have a `prefilter` method in the RParsePlugin, which is always called before colors are applied.

**Comment by XVilka** [comment]
`jmp 0xf000:0x4c5` is of course better. This is de-facto standard on representing these jumps.

**Comment by ret2libc** [comment]
@XVilka What about these instructions?

1. `cmp dword cs:[0x67b8], 0` (right now it is converted in `cmp dword 0xf000:[0x67b8], 0`)
2. `mov bl, byte cs:[edi - 0xeffd4]` (right now it is convert in `mov bl, byte 0xf000:[edi - 0xeffd4]`)

**Comment by radare** [comment]
that’s obviously a completely different thing

**Comment by ret2libc** [comment]
What do you mean @radare ? Can you explain more?

**Comment by radare** [comment]
cs is a segment register, which have unknown value at static time. 0xf000 is a value for the segment selector which is known at static analysis time. you shouldnt magically convert those values because the segment register value can differ at runtime

**Comment by ret2libc** [comment]
@radare that's right. The same is true for the jmp/call instructions, since cs value may be very different from 0xf000 (please notice that x86 uses the *cached segment base address* and not the direct segment value).

So are you suggesting we just use `cs:0xdead`? For calls and jmps, though, we still need to assume `cs` will be the same value, otherwise the address addresses of those instructions will be wrong.

**Comment by ret2libc** [comment]
@radare what do you think about this? we can choose whether to replace `cs:` with the static value where we loaded the binary (i've created a var called `asm.segsub`), but I think it's better this way more than having capstone resolving the linear address.

Right now segmentation is not well handled in radare2, but I think the disassembler (aka capstone) should be aware only of the EIP part of the segment and not the CS value. If you do agree, could you express your thoughts also on the capstone PR so that we can try to upstream this patch asap?

File: libr/parse/p/parse_x86_pseudo.c [hunk 2]
```diff
@@ -495,11 +499,45 @@ static bool varsub (RParse *p, RAnalFunction *f, ut64 addr, int oplen, char *dat
...
+	char *ptr, *rest;
+	if ((ptr = strstr (data, needle)) != NULL) {
+		rest = ptr + strlen (needle);
+		*ptr = '\0';
+
+		snprintf (str, len, "%s 0x%" PFMT64x ":%s", data, seg, rest);
+		return true;
+	}
+	return false;
+
```
Comment (ret2libc) [inline libr/parse/p/parse_x86_pseudo.c]: empty line

File: shlr/meson.build [hunk 3]
```diff
@@ -11,10 +11,10 @@ if not capstone_dep.found() or not get_option('use_sys_capstone')
     endif
 
     # NOTE: when you update CS_TIP or CS_BRA, also update them in shlr/Makefile
-    CS_TIP = '782ea67e17a391ca0d3faafdc365b335a1a8930a'
-    CS_BRA = 'master'
+    CS_TIP = ''
+    CS_BRA = 'fix-x86-16-pcrel'
 
-    capstone_git_user = 'aquynh'
+    capstone_git_user = 'ret2libc'
```
Comment (ret2libc) [inline shlr/meson.build]: fix this before merging and after capstone PR will be accepted

**Comment by ret2libc** [comment]
Ok, so: PR at https://github.com/aquynh/capstone/pull/1225 is not going to be merged, so we'll have to live without the `cs:` prefix. I think this is mostly fine. 

If we really want, we could still have `cs:` prefix or its value (e.g. `0xf000:`) prefixed in front of relative close/jump instructions, but we'd have to implement a bit more of code in the x86 parser, because we need to look at the detail of the disassembled instructions.

What do you guys think? 

I actually think it's fine even without the `cs:` prefix, look here:
bios bin

(we may need to fix the address in the graph, because it is showed in its linear form and not with `0xf000:d1c2`, but that's another story :) )

**Comment by radare** [comment]
seems like only insutrctions messing with absolute addreses are affected.

for example jmp, call, mov,.. but not with conditional jumps

so yep. the segment prefix may affect your example

**Comment by ret2libc** [comment]
@radare I'm not sure I understand your comments. Can you say more? Where do you see problems? How is the segment prefix going to improve/decrease the result?

**Comment by ret2libc** [comment]
@radare @XVilka ping

**Comment by ret2libc** [comment]
@radare ping

**Comment by radare** [comment]
did you checked the 6 XX? doesnt look good to me

**Comment by ret2libc** [comment]
ping @radare

File: libr/parse/p/parse_x86_pseudo.c [hunk 4]
```diff
@@ -495,11 +499,45 @@ static bool varsub (RParse *p, RAnalFunction *f, ut64 addr, int oplen, char *dat
 	return ret;
 }
 
+static bool do_replace_segment(ut64 seg, char *data, char *str, int len, const char *needle) {
+	char *ptr, *rest;
+	if ((ptr = strstr (data, needle)) != NULL) {
```
Comment (radare) [inline libr/parse/p/parse_x86_pseudo.c]: the != NULL is unnecessary

Comment (radare) [reply libr/parse/p/parse_x86_pseudo.c]: i prefer not to mix conditionals and assignments, and even less when the variable definition is not set

**Comment by radare** [comment]
maybe it will be good to have an option to choose if we want to resovle segment registers in the disasm filtering ?, but the overall idea lgtm

**Comment by ret2libc** [comment]
@radare thanks for the review. Given that near/relative jumps/calls do not have a `cs:` prefix, I was thinking about completely avoid the replace of `cs:` with its supposed value. About the option, there's already one in my code: `asm.segsub`. But are we sure we want it?

With `asm.segsub=false` you would have:
```
mov eax, cs:[0xbeef]
jmp 0xd1d2
```
With `asm.segsub=true`:
```
mov eax, 0xf000:[0xbeef]
jmp 0xd1d2
```

What about I just kill this filter in x86_parse and only apply the change to remove the capstone patch?

**Comment by radare** [comment]
i think its fine to have such option because it would be useful for static analysis, in debugging cs is always set, but if you want to analyze something outside the current scope you have to fake the cs value and restore it back later, or even if you want to rebase the CS a bunch of time to find the right value and compare the disasm, so i would keep that

**Comment by radare** [comment]
conflicting files, do u expect it to be ready before 1 week? otherwise maybe better to move for 3.1

**Comment by ret2libc** [comment]
I'll tell you in the next days whether I am able to make it for 3.0. I found some issues caused by not having the full linear address in the relative jumps/calls (e.g. you have a flag at `0xfdead` named `my_function`. You have a relative jump/call to `0xf000:0xdead`, thus being in the same segment it is shown as `jmp 0xdead`, without the `0xf000:` because the segment it's assumed to be the same. Unfortunately, the RParse is not able to translate `0xdead` into `my_function`, but it is able to translate `0xfdead` into it)

**Comment by radare** [comment]
Moving for 3.2? please resolve the conflicts and rebase

**Comment by ret2libc** [comment]
@p2mate what do you mean exactly? Can you make an example to help me understand better your point?

**Comment by p2mate** [comment]
> @p2mate what do you mean exactly? Can you make an example to help me understand better your point?

[doshello.zip](https://github.com/radare/radare2/files/2601354/doshello.zip)

if you look at this program, the call and jumps are near (in the same segment), so they should show as

0000:000f                 7404  je 0x15
0000:0013                 ebf5  jmp 0xa
and
0000:0021               e8dcff  call 0

the mov in the debugout subroutine has a segment override, so it should show as

0000:000a               268a07  mov al, byte es:[bx]

hope this helps.

**Comment by p2mate** [comment]
> > @p2mate what do you mean exactly? Can you make an example to help me understand better your point?

[doshello_far.zip](https://github.com/radare/radare2/files/2601648/doshello_far.zip)

Updated version with some more far (intersegment) pointers. The subroutine debugout is now in a different segment so the call in main looks like this in the disassembly:

    0000:0005           9a00000500  lcall 5:0

the jumps in debugout are still near so they still look like:

    0000:0063                 7404  je 0x69
    0000:0067                 ebf5  jmp 0x5e

the segment override is also the same:

    0000:005e               268a04  mov al, byte es:[si]

in DOS debug the program looks like this:

````
-u 0 l 16
2E21:0000 B8232E        MOV     AX,2E23                            
2E21:0003 8ED8          MOV     DS,AX                              
2E21:0005 9A0000262E    CALL    2E26:0000                          
2E21:000A BA0000        MOV     DX,0000                            
2E21:000D B409          MOV     AH,09                              
2E21:000F CD21          INT     21                                 
2E21:0011 B8004C        MOV     AX,4C00                            
2E21:0014 CD21          INT     21                                 
-u 2e26:0 l 1d
2E26:0000 50            PUSH    AX                                 
2E26:0001 56            PUSH    SI                                 
2E26:0002 52            PUSH    DX                                 
2E26:0003 BA0204        MOV     DX,0402                            
2E26:0006 BE0000        MOV     SI,0000                            
2E26:0009 B8242E        MOV     AX,2E24                            
2E26:000C 8EC0          MOV     ES,AX                              
2E26:000E 26            ES:                                        
2E26:000F 8A04          MOV     AL,[SI]                            
2E26:0011 3C24          CMP     AL,24                              
2E26:0013 7404          JZ      0019                               
2E26:0015 EE            OUT     DX,AL                              
2E26:0016 46            INC     SI                                 
2E26:0017 EBF5          JMP     000E                               
2E26:0019 5A            POP     DX                                 
2E26:001A 5E            POP     SI                                 
2E26:001B 58            POP     AX                                 
2E26:001C CB            RETF                                       
-d 2e23:0 l 10
2E23:0000  48 65 6C 6C 6F 20 77 6F-72 6C 64 21 0A 0D 24 CC   Hello world!..$.
-d 2e24:0 l 20
2E24:0000  48 65 6C 6C 6F 20 66 72-6F 6D 20 66 75 72 74 68   Hello from furth
2E24:0010  65 72 20 61 77 61 79 21-0A 0D 24 CC CC CC CC CC   er away!..$.....
````

**Comment by ret2libc** [comment]
@p2mate Ok I see what you mean and I do agree that is a good approach; that's what I was trying to achieve with this PR. However, there's a small catch related to how radare2 converts call/jmp addresses and similars to flags: it just parses the numbers in the disassembly string and check whethere there's a flag at that address. By using your solution, it is not able to find the right flags. I'm not saying this is not fixable, but it would involve some work.

Example:
```
2E21:000B 8A07          MOV     AL,[BX]                            
2E21:000D 3C24          CMP     AL,24                              
2E21:000F 7404          JZ      0015                               
2E21:0011 EE            OUT     DX,AL                              
2E21:0012 43            INC     BX                                 
2E21:0013 EBF5          JMP     000A    
end:                           
2E21:0015 5A            POP     DX                  
```
when looking at `jz 0x15`, RParse library see there is 0x15 and check in RFlags whethere there's a flag item at address 0x15. There is not (the flag `end` is indeed at address 0x2e225), so nothing is shown. To fix this RParse should be aware of the current address and know when a number is segment-relative or absolute (e.g. `jz 0x16` vs  `jz es:0xdeadbeef`).

**Comment by radare** [comment]
The rparse lib depends on arch. So just make the proper check to avoid sich conversions on 16 bit mode and it will be fine.

**Comment by zarszz** [comment]
i try too ..

**Comment by p2mate** [comment]
> > @p2mate Ok I see what you mean and I do agree that is a good approach; that's what I was trying to achieve with this PR. However, there's a small catch related to how radare2 converts call/jmp addresses and similars to flags: it just parses the numbers in the disassembly string and check whethere there's a flag at that address. By using your solution, it is not able to find the right flags. I'm not saying this is not fixable, but it would involve some work.
> > Example:
> > ```
> > 2E21:000B 8A07          MOV     AL,[BX]                            
> > 2E21:000D 3C24          CMP     AL,24                              
> > 2E21:000F 7404          JZ      0015                               
> > 2E21:0011 EE            OUT     DX,AL                              
> > 2E21:0012 43            INC     BX                                 
> > 2E21:0013 EBF5          JMP     000A    
> > end:                           
> > 2E21:0015 5A            POP     DX                  
> > ```
> > when looking at `jz 0x15`, RParse library see there is 0x15 and check in RFlags whethere there's a flag item at address 0x15. There is not (the flag `end` is indeed at address 0x2e225), so nothing is shown. To fix this RParse should be aware of the current address and know when a number is segment-relative or absolute (e.g. `jz 0x16` vs `jz es:0xdeadbeef`).
> 
> Indeed.. The parser needs to be segment aware. Maybe this will also solve the problem for lcall 5:0 in the second example. If you define a flag for 0x50, r2 won't notice that lcall 5:0 is a reference to it. Problem is that the segment:offset to linear address conversion only works for real mode 16bit binaries. For protected mode 16bit (eg NE binaries), the segments are selectors and the binary defines the number of segments and their size. The loader assigns the actual numbers and uses the relocs to patch them into the binary I think.



> @p2mate Ok I see what you mean and I do agree that is a good approach; that's what I was trying to achieve with this PR. However, there's a small catch related to how radare2 converts call/jmp addresses and similars to flags: it just parses the numbers in the disassembly string and check whethere there's a flag at that address. By using your solution, it is not able to find the right flags. I'm not saying this is not fixable, but it would involve some work.
> 
> Example:
> 
> ```
> 2E21:000B 8A07          MOV     AL,[BX]                            
> 2E21:000D 3C24          CMP     AL,24                              
> 2E21:000F 7404          JZ      0015                               
> 2E21:0011 EE            OUT     DX,AL                              
> 2E21:0012 43            INC     BX                                 
> 2E21:0013 EBF5          JMP     000A    
> end:                           
> 2E21:0015 5A            POP     DX                  
> ```
> 
> when looking at `jz 0x15`, RParse library see there is 0x15 and check in RFlags whethere there's a flag item at address 0x15. There is not (the flag `end` is indeed at address 0x2e225), so nothing is shown. To fix this RParse should be aware of the current address and know when a number is segment-relative or absolute (e.g. `jz 0x16` vs `jz es:0xdeadbeef`).

A similar problem would occur with eg:

mov al, es:[0x20] 

where you would need to know the value of es to be able to find the corresponding flag. I think we will need to have a feature to tell r2 what segment a given segment register points to at some point in the code. This would be similar to the assume statement in DOS assemblers.

**Comment by ret2libc** [comment]
Closing this as it's not going anywhere.

**Pull request closed without merging**